File: knowledge/public/templates/mobile-webview-handoff/ios/KyberionHandoffExporter.swift
import Foundation

enum KyberionHandoffError: Error {
    case disabled
}

protocol KyberionHandoffExporter {
    func exportWebViewSession(reason: String) throws -> URL
}

final class DefaultKyberionHandoffExporter: KyberionHandoffExporter {
    private let handoffEnabled: () -> Bool
    private let sessionReader: WebViewSessionReader
    private let storage: KyberionHandoffStorage
    private let appId: String

    init(
        handoffEnabled: @escaping () -> Bool,
        sessionReader: WebViewSessionReader,
        storage: KyberionHandoffStorage,
        appId: String = "example-ios-login-passkey"
    ) {
        self.handoffEnabled = handoffEnabled
        self.sessionReader = sessionReader
        self.storage = storage
        self.appId = appId
    }

    func exportWebViewSession(reason: String) throws -> URL {
        guard handoffEnabled() else {
            throw KyberionHandoffError.disabled
        }

        let state = try sessionReader.readCurrentState()
        let payload = KyberionWebviewSessionHandoff(
            targetURL: state.url.absoluteString,
            origin: state.origin,
            cookies: state.cookies,
            localStorage: state.localStorage,
            sessionStorage: state.sessionStorage,
            headers: state.headers.merging(["x-kyberion-reason": reason]) { _, new in new },
            appId: appId
        )
        return try storage.write(payload)
    }
}

File: knowledge/public/templates/mobile-webview-handoff/android/KyberionHandoffExporter.kt
package dev.kyberion.handoff

import java.io.File

interface KyberionHandoffExporter {
    fun exportWebViewSession(reason: String = "manual"): File
}

class DefaultKyberionHandoffExporter(
    private val handoffEnabled: () -> Boolean,
    private val stateReader: WebViewStateReader,
    private val storage: KyberionHandoffStorage,
    private val appId: String = "example-mobile-login-passkey"
) : KyberionHandoffExporter {

    override fun exportWebViewSession(reason: String): File {
        check(handoffEnabled()) { "Kyberion handoff is disabled" }

        val state = stateReader.readCurrentState()
        val payload = KyberionWebviewSessionHandoff(
            target_url = state.url,
            origin = state.origin,
            cookies = state.cookies,
            local_storage = state.localStorage,
            session_storage = state.sessionStorage,
            headers = state.headers + mapOf("x-kyberion-reason" to reason),
            source = mapOf(
                "platform" to "android",
                "app_id" to appId
            )
        )

        return storage.write(payload)
    }
}
